rollup/da_syncer/da/calldata_blob_source.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/blob_client"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/serrors"
+	"github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service"
+)
+
+const (
+	callDataBlobSourceFetchBlockRange  uint64 = 500
+	commitBatchEventName                      = "CommitBatch"
+	revertBatchEventName                      = "RevertBatch"
+	finalizeBatchEventName                    = "FinalizeBatch"
+	commitBatchMethodName                     = "commitBatch"
+	commitBatchWithBlobProofMethodName        = "commitBatchWithBlobProof"
+
+	// the length og method ID at the beginning of transaction data
+	methodIDLength = 4
+)
+
+var (
+	ErrSourceExhausted = errors.New("data source has been exhausted")
+)
+
+type CalldataBlobSource struct {
+	ctx                           context.Context
+	l1Client                      *rollup_sync_service.L1Client
+	blobClient                    blob_client.BlobClient
+	l1height                      uint64
+	scrollChainABI                *abi.ABI
+	l1CommitBatchEventSignature   common.Hash
+	l1RevertBatchEventSignature   common.Hash
+	l1FinalizeBatchEventSignature common.Hash
+	db                            ethdb.Database
+
+	l1Finalized uint64
+}
+
+func NewCalldataBlobSource(ctx context.Context, l1height uint64, l1Client *rollup_sync_service.L1Client, blobClient blob_client.BlobClient, db ethdb.Database) (*CalldataBlobSource, error) {
+	scrollChainABI, err := rollup_sync_service.ScrollChainMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get scroll chain abi: %w", err)
+	}
+	return &CalldataBlobSource{
+		ctx:                           ctx,
+		l1Client:                      l1Client,
+		blobClient:                    blobClient,
+		l1height:                      l1height,
+		scrollChainABI:                scrollChainABI,
+		l1CommitBatchEventSignature:   scrollChainABI.Events[commitBatchEventName].ID,
+		l1RevertBatchEventSignature:   scrollChainABI.Events[revertBatchEventName].ID,
+		l1FinalizeBatchEventSignature: scrollChainABI.Events[finalizeBatchEventName].ID,
+		db:                            db,
+	}, nil
+}
+
+func (ds *CalldataBlobSource) NextData() (Entries, error) {
+	var err error
+	to := ds.l1height + callDataBlobSourceFetchBlockRange
+
+	// If there's not enough finalized blocks to request up to, we need to query finalized block number.
+	// Otherwise, we know that there's more finalized blocks than we want to request up to
+	// -> no need to query finalized block number
+	if to > ds.l1Finalized {
+		ds.l1Finalized, err = ds.l1Client.GetLatestFinalizedBlockNumber()
+		if err != nil {
+			return nil, serrors.NewTemporaryError(fmt.Errorf("failed to query GetLatestFinalizedBlockNumber, error: %v", err))
+		}
+		// make sure we don't request more than finalized blocks
+		to = min(to, ds.l1Finalized)
+	}
+
+	if ds.l1height > to {
+		return nil, ErrSourceExhausted
```

Comment:
Would this only happen when "finalized batches are reverted"?

Reply:
or some bug cases? e.g., the fetcher missed some info but still updates the sync l1 height in db.

Reply:
This is due to the fact how the pipeline is called in `SyncingPipeline.mainLoop()`. We keep calling the pipeline and thus also querying L1 over and over until there's no more new finalized L1 blocks. At this point we introduce a delay with exponential backoff and wait until we try again. This happens basically forever. We return `serrors.EOFError` to indicate this to the pipeline: https://github.com/scroll-tech/go-ethereum/blob/00494d58cf3390cbf909a06ef3ef46b0587d6d90/rollup/da_syncer/syncing_pipeline.go#L174

Reply:
I see. It's L1's finalized block.